### PR TITLE
[Extend Governor APIs] Governor JSON Validator

### PR DESCRIPTION
**Pull request by bailinhe**
A customized JSON schema validator used to validate extension resources against the JSON schema defined in Extension Resource Definitions. This validator can be extended to meet the needs of the `governor-api`, as an example, a `unique-constrain` is implemented that allows extension developer to define fields that are unique within a given `ERD`

### CODE REVIEW

File: pkg/jsonschema/unique_constraint.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+	qms := []qm.QueryMod{}
+
+	if s.ResourceID != nil {
+		qms = append(qms, qm.Where("id != ?", *s.ResourceID))
+	}
+
+	for k, value := range mappedValue {
+		// Convert the value to string
+		v := fmt.Sprint(value)
```
Comment (fishnix) [inline pkg/jsonschema/unique_constraint.go:69]: why convert the value to a string here? should be it a type assertion? should it be for any type?

File: pkg/jsonschema/unique_constraint.go [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+	return &UniqueConstraintSchema{resultUniqueFields, uc.ERD, uc.ResourceID, uc.ctx, uc.db}, nil
+}
+
+// Checks if the provided field type is valid for unique constraints
+func isValidType(fieldType string) bool {
+	_, ok := map[string]bool{
+		"string": true, "number": true, "integer": true, "boolean": true,
+	}[fieldType]
+
+	return ok
```
Comment (fishnix) [inline pkg/jsonschema/unique_constraint.go:199]: Is this any better than just returning from a conditional?

```suggestion
	return fieldType == "string" || fieldType == "number" || fieldType == "integer" || fieldType ==  "boolean"
```

File: pkg/jsonschema/unique_constraint.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+// helper function to assert string slice type
+func assertStringSlice(value interface{}) ([]string, error) {
```
Comment (fishnix) [inline pkg/jsonschema/unique_constraint.go:203]: test for this?

**Review by fishnix — COMMENTED** [review]
Just a few comments - will this be imported by addons? If not should it be in `internal` ?

File: pkg/jsonschema/unique_constraint.go (hunk 2 above)
Comment (bailinhe) [reply pkg/jsonschema/unique_constraint.go:199]: yes, I was originally thinking to use a set, then realize go doesn't have sets, ended up over thinking this

File: pkg/jsonschema/unique_constraint.go (hunk 1 above)
Comment (bailinhe) [reply pkg/jsonschema/unique_constraint.go:69]: the `Compile` function will ensure the types are not `object` or `array` (by using the `isValidType` function below). Here covert the value to string so I can use `"%s"` in the `Sprintf` statement below

**Comment by bailinhe** [comment]
> Just a few comments - will this be imported by addons? If not should it be in `internal` ?

I hope this can be used for addons to self-validate the schemas provided by the developer, make sure they are compilable, perhaps as one of the tests in the addons

File: pkg/jsonschema/unique_constraint.go (hunk 3 above)
Comment (bailinhe) [reply pkg/jsonschema/unique_constraint.go:198]: added new test [here](https://github.com/metal-toolbox/governor-api/pull/74/files#diff-52d0f28dc80d9ae67dac99fe3a7b186bd8260eb821250c08cb9976372b1573adR270-R313)

**Review by fishnix — APPROVED** [review]

**Pull request merged**